### DISCUSSION TITLE
milvus1.1.1，具有多个partition的集合边写入边查询时，查询的rt会有一定波动。

### DISCUSSION THREAD

**Discussion by SuoSiFire** [milvus-io/milvus#16171] [Q&A and General discussion] [answered]
milvus1.1.1，具有多个partition的集合边写入边查询时，查询的rt会有一定波动。
场景是写节点只有一台，读节点有多台，不同的读节点会加载不同的partition。 每个partition有多个机器加载。在查询的时候，最后一个partition的rt波动会比加大，当分片较多时，rt会从300ms 波动到 1 - 2s。 
 这是不是由于最后一个segment的索引还没有建立，这一部分是暴力查询的，并且当最后一个segment构建索引和读节点加载索引的过程中，也会有一定的延迟导致的？ 
请教下这种rt波动的具体原因和规避方案。

**Comment by yhmo** [comment, marked as answer]
先调用create_index()再持续调用insert(),这种方式会在后台自定建索引，当某个segment小于index_file_size的大小时，不会被拿去建索引，而是会参与合并，几个小分片合成一个大分片，当分片大于index_file_size时，就会被拿去建索引。所以最后那个segment在不断变大的过程中，查询性能不断下降，直到当它被拿去建好索引之后，查询性能立刻恢复最优。所以在这种情况下，查询性能的曲线形状大致如下：
```
    /|    /|    /|    /|    /|    /|    /|    /|    /|
  /  |  /  |  /  |  /  |  /  |  /  |  /  |  /  |  /  |
/    |/    |/    |/    |/    |/    |/    |/    |/    |
```

使用小一点的index_file_size值，会使得这个锯齿型的锯齿变小，使用大一点的值，会使得锯齿变大。但只要是这种streaming的流式数据场景，就无法杜绝这个锯齿型性能的现象。

**Reply by SuoSiFire** [reply]
感谢回复 @yhmo ，性能曲线的确是这样的，并且 999线 可能超过20s，这在业务场景内会导致请求因为超时而失败，进而引起重试，导致请求失败率会在短时间内比较高。  

我尝试过调整index_file_size,但是过小的index_file_size会导致整体的查询性能变差（可能是segment太多导致的？）。听您的描述主要问题应该是最后的segment暴力查询会导致整体的rt波动。 那么我在上层弄一个临时集合的概念，每次查询会同时查询全量集合和临时集合， 临时集合会（定期/定量）合并到全量集合里，这样临时集合的index_file_size就可以设置的比较小，并且可以设置更加合适的索引类型。您觉得这样可行吗？ 期待回复。

**Reply by SuoSiFire** [reply]
另外还有个问题是 https://github.com/milvus-io/milvus/issues/6684 我在这里回复的，使用修改后的getEntityById去获取10w条记录，需要10min的时间，请问是使用方式的问题，还是性能就是这样的？

**Reply by yhmo** [reply]
数据总量如果在千万级，单次查询一般不会有20秒那么久，看看是不是cache_size设置的太紧。
临时集合的数据如果能和主集合的数据无缝过渡的话，是有可能把性能曲线的锯齿钝化。但要做到无缝过渡，操作上似乎不具备可行性。

getEntityById实现上是根据id定位到某个segemnt数据文件的某个磁盘位置，读取一条向量的一小段数据。之所以是读取磁盘而不是读取内存是因为：除了IVFFLAT索引之外，其他的索引，要么难以定位向量位置，要么向量被做了编码变得跟原始数据不同，所以无法获取向量的原始数据。建了索引之后，读入内存的是索引而不是向量的原始数据，所以只能通过读取磁盘来获取。 #6684对批量获取的请求做了一些并行处理，但瓶颈仍然在读磁盘上。10w条记录，如果你分批地调用getEntityById，每批1千条的话调用100次，应该会比调用10w次getEntityById要快。

**Reply by SuoSiFire** [reply]
感谢回复。


数据总量目前大概7亿左右，单个partition 5kw大小，分为15个partition。256维向量，ivfsq8索引，index_file_size 设置的是2048，nlist为4096.   单机内存56G，cache size目前设置的是40G。 
麻烦帮忙看下有哪些参数可以调整吗？ 


我应该就是每批1k条调用的，因为使用getEntityById接口时，貌似有个批量上限。那看起来性能就是这样了？

**Reply by yhmo** [reply]
7亿数据256维有700GB，全部建好ivfsq8索引后也有200G左右大小，所以40G的cache_size肯定是不足以把全表放在内存里搜的，除非你只搜索一个partition。
单个分片2GB，每个分片大约200万条，根据4 × sqrt(n)的经验来设nlist的话，4096也合适，也不用调整了。要么就只能把搜索参数nprobe设小一点了，或者如果对召回率要求不高的话考虑使用ivfpq索引。
getEntityById的性能基本就这样了，实际上就是每次调用找到1000条向量的磁盘位置，分1000次读取1KB数据(单条向量大小)，因为不是连续读取所以挺慢的，尤其对于非本地磁盘存储来说。

**Reply by SuoSiFire** [reply]
好的，感谢回复。 目前确实是通过控制请求来保证单机只加载一个partition的。

**Reply by SuoSiFire** [reply]
@yhmo  打扰一下，我又观察了下我这边的性能曲线，整体的rt和您的描述相似，但单独看一段时间内的rt，是有比较大的波动的，我这边的rt曲线大致像这样
![rt](https://user-images.githubusercontent.com/90120870/161281640-982b474c-aefb-4c7b-a600-4623409741cf.png)

看最近3h的监控情况，3h内平均rt为800ms，但是平均rt每隔5min就会出现一个峰值，峰值大概在5s左右，持续1min。 请问这可能和什么有关，我这边几个实时写入+查询的集合都有这么大的rt抖动问题。  这可能和cache_size大小有关，或者一些配置参数有关吗？ 

还有auto_flush_interval这个参数目前设置的是1s，是否合理呢？  

期待回复！

**Reply by yhmo** [reply]
存储使用的是共享存储(比如NAS)还是本地磁盘？有增量数据的时候，查询请求要把增量数据产生的新segment加载到内存，需要从存储系统读数据。同时insert不断地把增量数据送进来，每次flush都会产生一个segment，并通过IO写入。同时后台不断地把新产生的小分片从存储系统读上来，合并成大分片，每次合并完成也会通过IO写入大分片。

共享存储带宽是有限的，在后台正在写入一个大分片或者读取若干个的时候，查询请求可能会在加载新分片的时候变得很慢。

比如，刚开始时，insert进来的数据生成了两个1MB的小分片，后台合并成一个2MB的分片，查询请求把2MB的新分片加载进来查询，这个耗时很短，因为2MB的IO读写不会有什么问题。随着插入数据不断进来，新分片变得越来越大，最差的时候是1GB+1GB=2GB的时候，这时要写入2GB的文件，查询请求把2GB的文件读上来，所以会很慢。

如果把index_file_size设为1GB，观察性能，我估计峰值应该会小一半左右。
auto_flush_interval设为1s，是希望1s之前进来的数据能被搜索到(因为不flush的数据是不能搜索到的)，这个值设大了，只会影响数据的可见性，对最终要合并的大小没影响，最终合并的大小取决于index_file_size

**Reply by yhmo** [reply]
另外，两个波峰之间的时间间隔主要跟插入的速度相关。插入速度快，波峰之间的距离就小，插入速度慢，波峰之间距离就长。从你的图上看，可能你每5 ~ 6分钟insert进来的数据量大约就是2GB(index_file_size)大小

**Reply by SuoSiFire** [reply]
1. 存储使用的是cfs共享文件系统，确实会比本地磁盘慢些。
2. 每5~6分钟insert进来的数据量并没有这么大，插入的速度大概50qps,5分钟差不多也就15M的数据量。所以不明白这个波动是后台的什么操作导致的。

**Reply by SuoSiFire** [reply]
目前每个partition可能会被8+个读节点加载，是不是可能在查询请求过来时恰好后台在合并操作，每个读节点同时加载新分片，导致这段rt突增呢？

**Reply by yhmo** [reply]
可能是因为数据被分到8个分区，每当某个分区的数据达到合并的条件时，后台合并产生一个新分片，新分片的存盘和加载都会产生耗时。
在1.x里，合并的逻辑参考这篇：https://zhuanlan.zhihu.com/p/148106312
一般情况下，为了减少“写放大”效果，只会把较小的分片拿来合并，但不能一直把小分片一直搁着，要定期强制把小分片合进大分片里，这个定期强制的时间就是5分钟，所以才会产生你看到的效果。

举例：
假设已经有一个500M大小的分片存在，然后你第1分钟插入3MB数据，第2分钟又插入3MB数据，这时milvus只会把两个3MB的小分片合并成一个6MB的新分片，这个新分片对性能的影响很小，然后第3分钟又插入3MB，第4分钟又3MB，milvus把两个3MB合并成另一个6MB，接着两个6MB合成12MB，也不算大，对性能影响很小。但到了第5分钟，强制合并的时间到了，milvus把500MB的分片跟这个12MB的分片合并成一个512MB的新分片，然后查询请求要加载这个新分片，这时就产生了512MB的写操作和512MB的读操作，对IO的压力较大，于是形成了一个波峰。所以，在存在一个较大分片(小于index_file_size)时，每次强制合并都会产生较大的IO操作，从而形成一个个的波峰。

这个5分钟强制合并的间隔是写死在代码里的，不能改动。

**Reply by SuoSiFire** [reply]
感谢回复，听上去这个描述比较符合我们的性能曲线。 我这边是自己编译的milvus源码，我应该可以通过调整这个强制合并时间间隔来减少波峰的出现频率吧。 听上去调大这个强制合并时间只会影响到小分片合并进大分片的时间，只要不调整的过大导致小文件太多就好了？ 

可以给个调整的建议吗，我这边试试。

**Reply by yhmo** [reply]
在这个源文件里：milvus/core/src/db/merge/MergeLayeredStrategy.cpp
`const int64_t FORCE_MERGE_THREASHOLD = 300;  // force merge files older than this time(in second)`
改变这个值就能观察到变化。

合并是按大小分层的，index_file_size=2GB的时候，会根据4Mb,16MB,64MB,256MB,1GB这几个档位分别合并，比如3MB的跟1MB的合并，100MB的跟200MB的合并，不会把1MB和100MB合在一起。如果把这个间隔时间调成很长，观察产生的分片，会发现大致每个档位里会产生一个分片。

小档位产生合并时，IO压力小，对性能影响小。大档位产生合并时，还是会产生一个性能上的波峰。比如100MB和200MB合并，产生300MB的分片，这时就进入了256MB这个档位，然后如果之前已存在一个280MB的分片的话，就会跟300MB的合并，产生一个580MB的分片。

把时间间隔变长，波峰的数量会减少。

**Reply by SuoSiFire** [reply]
感谢回复，我这边尝试下，看看会不会对整体性能有一定提升。

**Reply by SuoSiFire** [reply]
@yhmo  按照上面说的修改FORCE_MERGE_THREASHOLD 到了30分钟，rt波动的问题解决了。 但是修改后mysql的thread running指标相较于之前上涨很多，想了解下这个改动会造成某种类型的查询执行频率上升吗？ 或者会造成某种类型的查询扫描行数变多吗。

**Reply by yhmo** [reply]
这个FORCE_MERGE_THREASHOLD只是影响segment的数量，之前每5分钟强制合并一次，小于index_file_size的segment数量就1 ～ 2个，改成30分钟之后，小于index_file_size的segment数量可能会有4 ～ 5 个。mysql里记录的那个TableFiles表的记录会多几条，但应该没什么影响。

**Comment by SuoSiFire** [comment]
@yhmo  又提出了一些问题，麻烦有空帮忙看下
